src/aspire/utils/blk_diag_matrix.py

```diff
@@ -0,0 +1,700 @@
+    #   because we implement them here ourselves.
+    # This is a more np current implementation of __array_priority__
+    #   operator precedence schedule.
+    # Mainly this effects rmul, radd, rsub eg:
+    #   blk_y = scalar_a + ( scalar_b * blk_x)
+    __array_ufunc__ = None
+
+    def __init__(self, nblocks, partition=None, dtype=np.float64):
+        """
+        Instantiate a BlkDiagMatrix.
+
+        :param nblocks: Number of diagonal matrix blocks
+        :param partition: Optionally supply the matrix block partition
+         in the form of a `nblock`-element list storing all shapes of
+         diagonal matrix blocks, where `partition[i]` corresponds to
+         the shape (number of rows and columns) of the `i` matrix block.
+        :return BlkDiagMatrix instance
+
+        """
+
+        self.nblocks = nblocks
+        self.dtype = np.dtype(dtype)
+        self.data = [[]] * nblocks
+        self._cached_blk_sizes = None
+        if partition:
+            self._cached_blk_sizes = np.array(partition)
+            assert self._cached_blk_sizes.shape[1] == 2
+
+    def __repr__(self):
+        """String represention describing instance """
+        return "BlkDiagMatrix({}, {})".format(
+            repr(self.nblocks), repr(self.dtype))
+
+    def copy(self):
+        """
+        Returns new BlkDiagMatrix which is a copy of `self`.
+        :return BlkDiagMatrix like self
+        """
+        return BlkDiagMatrix.from_blk_diag(self.data)
+
+    # Manually overload list methods,
+    #   This is just for syntax which allows us to reference self[i] etc
+    #     instead of writing self.data all the time. You may use either.
+    def __getitem__(self, key):
+        """ Convenience wrapper, getter on self.data """
+        return self.data[key]
+
+    def __setitem__(self, key, value):
+        """ Convenience wrapper, setter on self.data """
+        self.data[key] = value
+
+    def __len__(self):
+        """ Convenience function for getting nblocks """
+        return self.nblocks
+
+    def _is_scalar_type(self, x):
+        """ Internal helper function checking scalar-ness for elementwise ops
+
+        :return: bool
+        """
+        # Implemented as a function in case we need to modify it in future...
+        return np.isscalar(x)
+
+    @staticmethod
+    def zeros(blk_partition, dtype=np.float64):
+        """
+        Build a BlkDiagMatrix zeros matrix
+
+        :param blk_partition: The matrix block partition in the form of a
+        K-element list storing all shapes of K diagonal matrix blocks,
+        where `blk_partition[i]` corresponds to the shape (number of rows and
+        columns) of the `i` diagonal matrix block.
+        :param dtype: The data type to set precision of diagonal matrix block.
+        :return: A BlkDiagMatrix matrix consisting of `K` zero blocks.
+        """
+
+        n = len(blk_partition)
+        A = BlkDiagMatrix(n, dtype=dtype)
```

Comment:
It doesn't need it, but we can.  It will only have the effect of caching the sizes now instead of first touch.

I actually struggled a little bit considering what the default constructor should do and settled on what I felt was an `empty` equivalent.  Open to change.

Reply:
The `__init__` method now takes the partition instead.